contracts/GenArt721CoreV3_Engine_Flex.sol

```diff
@@ -0,0 +1,2278 @@
+     */
+    function updateArweaveGateway(string calldata _gateway) public {
+        _onlyAdminACL(this.updateArweaveGateway.selector);
+        preferredArweaveGateway = _gateway;
+        emit GatewayUpdated(ExternalAssetDependencyType.ARWEAVE, _gateway);
+    }
+
+    /**
+     * @notice Locks external asset dependencies for project `_projectId`.
+     */
+    function lockProjectExternalAssetDependencies(uint256 _projectId) external {
+        _onlyUnlockedProjectExternalAssetDependencies(_projectId);
+        _onlyArtistOrAdminACL(
+            _projectId,
+            this.lockProjectExternalAssetDependencies.selector
+        );
+        projects[_projectId].externalAssetDependenciesLocked = true;
+        emit ProjectExternalAssetDependenciesLocked(_projectId);
+    }
+
+    /**
+     * @notice Returns external asset dependency for project `_projectId` at index `_index`.
+     * If the dependencyType is ONCHAIN, the `data` field will contain the extrated bytecode data and `cid`
+     * will be an empty string. Conversly, for any other dependencyType, the `data` field will be an empty string
+     * and the `bytecodeAddress` will point to the zero address.
+     */
+    function projectExternalAssetDependencyByIndex(
+        uint256 _projectId,
+        uint256 _index
+    ) public view returns (ExternalAssetDependencyWithData memory) {
+        ExternalAssetDependencyType _dependencyType = projects[_projectId]
+            .externalAssetDependencies[_index]
+            .dependencyType;
+        address _bytecodeAddress = projects[_projectId]
+            .externalAssetDependencies[_index]
+            .bytecodeAddress;
+
+        return
+            ExternalAssetDependencyWithData({
+                dependencyType: _dependencyType,
+                cid: projects[_projectId].externalAssetDependencies[_index].cid,
+                bytecodeAddress: _bytecodeAddress,
+                data: (_dependencyType == ExternalAssetDependencyType.ONCHAIN)
+                    ? _bytecodeAddress.readFromBytecode()
+                    : ""
+            });
+    }
+
+    /**
+     * @notice Returns external asset dependency count for project `_projectId` at index `_index`.
+     */
+    function projectExternalAssetDependencyCount(
+        uint256 _projectId
+    ) public view returns (uint256) {
+        return uint256(projects[_projectId].externalAssetDependencyCount);
+    }
+
+    /**
+     * @notice Updates external asset dependency for project `_projectId`.
+     * @param _projectId Project to be updated.
+     * @param _index Asset index.
+     * @param _cidOrData Asset cid (Content identifier) or data string to be translated into bytecode.
+     * @param _dependencyType Asset dependency type.
+     *  0 - IPFS
+     *  1 - ARWEAVE
+     *  2 - ONCHAIN
+     */
+    function updateProjectExternalAssetDependency(
+        uint256 _projectId,
+        uint256 _index,
+        string memory _cidOrData,
```

Comment:
thinking this should be a generic bytes, with an update our bytecode storage library to handle bytes... let's chat sync about this in some more detail though!

Reply:
per discussion, this seems good as-is 👍 happy to discuss in more detail if we learn anything new